### PR TITLE
Fix an issue in `TimeSeriesProperty`

### LINKED ISSUE #18944: Fix crash with OFFSPEC data
Mantid crashes with some `OFFSPEC` runs

### Expected behavior

It shouldn't crash

### Actual behavior

It crashes

### Steps to reproduce the behavior

Run this script:

```
ws_1 = LoadISISNexus(Filename=r'OFFSPEC00038084.nxs')
ws_2 = LoadISISNexus(Filename=r'OFFSPEC00038087.nxs')
dbp = Plus(LHSWorkspace='ws_1', RHSWorkspace='ws_2')
```

### Platforms affected

Windows, debug mode

### PR DESCRIPTION
**Pull request by raquelalvarezbanos**
Description of work.

**To test:**

- This has to be tested on Windows, debug mode.
- Run the script in the issue description, Mantid shouldn't crash.
- Make sure that the new unit test fails without this fix.
- I could have added a similar check for the property values before [this](https://github.com/mantidproject/mantid/blob/291847a5e65364b6933fcf325f91d15932b99f4c/Framework/Kernel/src/TimeSeriesProperty.cpp#L191) line, but I think it is unnecessary. Make sure you agree with me.

Fixes #18944.

No need to mention this in the release notes, as the bug only happens on Windows in debug mode.

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [ ] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?

- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/Kernel/src/TimeSeriesProperty.cpp [hunk 1]
```diff
@@ -176,6 +176,9 @@ operator==(const TimeSeriesProperty<TYPE> &right) const {
   { // so vectors can go out of scope
     std::vector<DateAndTime> lhsTimes = this->timesAsVector();
     std::vector<DateAndTime> rhsTimes = right.timesAsVector();
+    if (lhsTimes.size() != rhsTimes.size()) {
```
Comment (peterfpeterson) [inline Framework/Kernel/src/TimeSeriesProperty.cpp]: Shouldn't you use  `TimeSeriesProperty::size()` or `TimeSeriesProperty::realSize()` and copy the vectors if they aren't equal?

**Review by peterfpeterson — CHANGES_REQUESTED** [review]

File: Framework/Kernel/src/TimeSeriesProperty.cpp (hunk 1 above)
Comment (DimitarTasev) [inline Framework/Kernel/src/TimeSeriesProperty.cpp]: :+1: maybe extend the `if (this->m_size != right.m_size) ` above with an ` || this->realSize() != right.realSize()`, then the check will be applied for both below

Comment (peterfpeterson) [inline Framework/Kernel/src/TimeSeriesProperty.cpp]: The greatest effect would be had by putting this as an `if` on line 176.

Comment (DimitarTasev) [inline Framework/Kernel/src/TimeSeriesProperty.cpp]: @peterfpeterson Could you tell me why an additional `if` is preferred to extending an existing one?

**Comment by DimitarTasev** [comment]
I've tested without the fix works and mantid longer crashes. After applying the fix there is no crash.
I've ran the unit tests without the fix and the new test fails. After applying the fix it passes successfully.
If @peterfpeterson doesn't have any comments I'm happy to :shipit:

File: Framework/Kernel/src/TimeSeriesProperty.cpp (hunk 1 above)
Comment (raquelalvarezbanos) [inline Framework/Kernel/src/TimeSeriesProperty.cpp]: Shouldn't they be the same?

File: Framework/Kernel/src/TimeSeriesProperty.cpp [hunk 2]
```diff
@@ -173,12 +173,13 @@ operator==(const TimeSeriesProperty<TYPE> &right) const {
     return false;
   }
 
+  if (this->realSize() != right.realSize()) {
+    return false;
+  }
+
```
Comment (peterfpeterson) [inline Framework/Kernel/src/TimeSeriesProperty.cpp]: add the word `else` here and I'm happy

**Comment by peterfpeterson** [comment]
I think a clearer version of the code (with added `const` below) is:
```c
if (this->realSize() != right.realSize()) {
     return false;
} else {
     const std::vector<DateAndTime> lhsTimes = this->timesAsVector();
      const std::vector<DateAndTime> rhsTimes = right.timesAsVector();
     if (!std::equal(lhsTimes.begin(), lhsTimes.end(), rhsTimes.begin())) {
       return false;
     }

     const std::vector<TYPE> lhsValues = this->valuesAsVector();
     const std::vector<TYPE> rhsValues = right.valuesAsVector();
     if (!std::equal(lhsValues.begin(), lhsValues.end(), rhsValues.begin())) {
       return false;
     }
}

return true;
```
I don't think the `const` buys anything since those functions return copies of the vectors. Without checking against a compiler, you can probably declare the various variables as `const auto` as well, but I don't think it really adds.

**Review by peterfpeterson — APPROVED** [review]

**Comment by martyngigg** [comment]
Tests are the usual unrelated failures.

**Pull request merged**